services/functions/linkdex-cid.ts

```diff
@@ -0,0 +1,71 @@
+import { APIGatewayProxyHandlerV2, APIGatewayProxyStructuredResultV2 } from "aws-lambda"
+import { S3Client, _Object } from "@aws-sdk/client-s3"
+import { CID } from 'multiformats/cid'
+import { DagStructure, LinkIndexer } from 'linkdex'
+import pMap from 'p-map'
+import { listCars, indexCar } from './lib/s3.js'
+import { response } from './lib/api.js'
+
+export const handler: APIGatewayProxyHandlerV2 = async (event) => {
+  const cid = event.pathParameters?.cid ?? ''
+  const bucket = process.env.BUCKET_NAME ?? ''
+  try {
+    return await getLinkdexReportForCid(cid, bucket, new S3Client({}))
+  } catch (err: any) {
+    const details = err?.message ?? err
+    return response({ error: { details }}, { statusCode: 500 })
+  }
+}
+
+/**
+ * Try and find a set of CARs that contain a complete DAG for a given CID.
+ * For dotStorage CARs are grouped by userid with keys like `raw/<cid>/<userid>/<hash>.car`.
+ * We find all the CARs for a given cid by prefix, group them by userid, 
+ * and then sort by count of CARs, under the assumptions that the upload with the most CARs
+ * is likely to contain the full DAG.
+ */
+export async function getLinkdexReportForCid (cid: string, bucket: string, s3: S3Client) {
+  if (!cid) {
+    return response({ error: { details: 'cid path param is required' }}, { statusCode: 400 })
+  }
+  if (!CID.parse(cid)) {
+    return response({ error: { details: 'cid is invalid' }}, { statusCode: 400 })
+  }
+  const prefix = `raw/${cid}/`
```

Comment:
Not blocking, but we should consider CARs in `complete/${cid}`. For examples, CARs from [backup](https://github.com/web3-storage/backup/blob/main/index.js#L133) are written there.